Service-Client and Microservice support for OASP4J

hohwille opened the pull request:
This is my initial proposal for supporting service clients (https://github.com/oasp/oasp4j/issues/358) and microservices (https://github.com/devonfw/devon/issues/65).

The solution is already very flexible and customizable.

I still see a lot of room to discuss and improve:
* [x] soap client is not yet supported/implemented
* [x] should we split/separate spring-boot auto configs (from `oasp4j-cxf-server` split off `oasp4j-cxf-server-rest` and `oasp4j-cxf-server-soap` and same for client) - results in JARs with 1 class each but more flexibility via dependency control and "automagic".
* [ ] add more logging
* [ ] add more JUnit tests
* [x] use it consistently across `oasp4j-sample-core`
* [ ] async client support
* [ ] do we want to build in resilence or follow sidecar approach?
* [ ] also extend oasp4j-security with OAuth/JWT support to work smooth with `ServiceHeaderCustomizerOAuth`.

For me the most important part and question is:
* [ ] does the current API design fit and is it reliable, sane and stable?
* [ ] we can always extend and improve in later releases (we have a flexible and working implementation right now, netflix support, etc. can now be added by projects who are using netflix stack).

hohwille commented:
I am also working on support for automatic @Inject support for spring. However, this is more tricky as I hoped:
https://stackoverflow.com/questions/39348010/how-to-create-a-bean-by-type-in-spring
For CXF this should work out well as clients are thread-safe:
https://stackoverflow.com/questions/13136167/are-cxf-client-proxies-thread-safe

See also discussion here:
https://github.com/oasp/oasp4j/issues/358

amarinso commented:
@hohwille playing a bit with Spring for scanning annotations and injecting beans I created this simple sample testcase at [https://github.com/amarinso/spring-dynamic-bean](https://github.com/amarinso/spring-dynamic-bean)

Do you think it could be useful for our case?

I haven't fiddle with Spring internals in a long time so it may not be the best option... but was fun to try :-)

amarinso commented on samples/core/src/test/java/io/oasp/gastronomy/restaurant/salesmanagement/service/impl/rest/SalesmanagementHttpRestServiceTest.java:
```diff
@@ -61,6 +63,9 @@
   private SalesmanagementRestServiceTestHelper helper;
 
   @Inject
+  private ServiceClientFactory serviceClientFactory;
```
Could the injection use Generics?
```
@Inject
private ServiceClientFactory<SalesmanagementRestService> serviceClientFactory;
```

So later we don't need to specify the class?
`this.service = this.serviceClientFactory.create()`

hohwille replied:
The design of `ServiceClientFactory` is intended the way it is in order to be generic and flexible. Using generics to get information from it is very tricky due to erasure so you can not actually retrieve the type except via wild reflection tricks where you would have to know the name of the field and still a lot of complex stuff would need to be done. I went through all this hell when implementing support for it in m-m-m:
https://m-m-m.github.io/maven/apidocs/net/sf/mmm/util/reflect/api/GenericType.html
So we could create a new interface for this semantic with a non-arg `create()` method and use it as you propose. However, this would require that you get an individual instance injected and that depends on reflective implementation of the underlying bean (that might already be weaved with AspectJ). To sum up: Nice idea but would require much magic, will be fragile and might break with the next spring release, and in the end is not even shorter (ok, you actually save 3 characters in total).

amarinso replied:
My original intent was not to save typing but to delegate object creation to the IoC, whenever I see some code creating objects that could be injected seems a bit odd.

Isn't this more or less the same as Spring Data does with repositories?

I will review your code in order to get more insight into this

hohwille commented:
> @hohwille playing a bit with Spring for scanning annotations and injecting beans I created this simple sample testcase at https://github.com/amarinso/spring-dynamic-bean
> Do you think it could be useful for our case?

@amarinso great. Thanks for sharing. This already brought me closer to a working solution. I integrated this locally in my fork and made some progress. However, spring really has some flaws as it seems:
* classpath scanning of spring does not seem to do any caching. This approach slows down the startup extraordinary (nothing to wonder if they provide classname as string even though it the Class must already have bean loaded before for filtering so in the end spring has to do `Class.forName` sereral times on the same classname - IMHO bad design).
* it can get faster if you define the base package(s) to scan (instead of root). However, how could we know them. I tried to solve this via configuration but...
* in order to get `registerBeanDefinitions` from `ImportBeanDefinitionRegistrar` invoked, you need to define an `@Import` on it as you did in your demo. However, then this method gets invoked before anything is injected. So no configuration or dependent beans at this point.
* we could not use `ImportBeanDefinitionRegistrar` and instead use a regular initializer via `@PostConstruct`. But maybe then the registration of the beans might happen too late. Who knows for sure.
* I wanted to test the result in our sample core but the problem is that we already have implementations of the service interface as spring beans to provide the service. Therefore spring will inject the service implementation instead of the client stub.

So as you can see this is still experimental work in progress and I am unsure if I can actually get this solved properly. Maybe it is better to create a feature request in spring framework for this feature: https://stackoverflow.com/questions/44495708/dynamic-spring-factory-to-create-a-bean-requested-by-its-given-type

hohwille commented:
In case someone wants to play around with the state I have created:
```
package io.oasp.module.service.common.impl;

import java.lang.annotation.Annotation;

import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.BeanFactoryAware;
import org.springframework.beans.factory.FactoryBean;
import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider;
import org.springframework.context.annotation.ImportBeanDefinitionRegistrar;
import org.springframework.core.type.AnnotationMetadata;
import org.springframework.core.type.filter.AnnotationTypeFilter;

import io.oasp.module.service.common.api.client.ServiceClientFactory;

/**
 * Implementation of {@link ImportBeanDefinitionRegistrar} that scans the class-path for service interfaces and
 * automatically registers {@link FactoryBean}s for them based on {@link ServiceClientFactory} so you can simply
 * {@literal @}{@link Inject} service interfaces and get a fully configured client proxy to access the actual remote
 * service over the network.
 */
public class ServiceClientSpringFactory implements ImportBeanDefinitionRegistrar, BeanFactoryAware {

  /** Logger instance. */
  private static final Logger LOG = LoggerFactory.getLogger(ServiceClientSpringFactory.class);

  private ServiceClientFactory serviceClientFactory;

  private ConfigurableListableBeanFactory beanFactory;

  // set to empty value to disable entirely
  @Value("${service.api.packages:*}")
  private String packages;

  @Value("${service.api.annotations:javax.ws.rs.Path,javax.jws.WebService}")
  private String annotationQualifiedNames;

  /**
   * The constructor.
   */
  public ServiceClientSpringFactory() {
    super();
  }

  @Override
  public void setBeanFactory(BeanFactory beanFactory) throws BeansException {

    this.beanFactory = (ConfigurableListableBeanFactory) beanFactory;
  }

  @Override
  public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata, BeanDefinitionRegistry registry) {

    this.serviceClientFactory = this.beanFactory.getBean(ServiceClientFactory.class);
    if (this.serviceClientFactory == null) {
      LOG.error("ServiceClientFactory not available. Injection of client stubs via spring not available!");
      return;
    }
    ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(false) {
      @Override
      protected boolean isCandidateComponent(AnnotatedBeanDefinition beanDefinition) {

        // scan for interfaces...
        return (!beanDefinition.getMetadata().isConcrete() && beanDefinition.getMetadata().isIndependent());
      }
    };
    if (this.annotationQualifiedNames == null) {
      this.annotationQualifiedNames = "javax.ws.rs.Path,javax.jws.WebService";
    }
    String[] annotationFqns = this.annotationQualifiedNames.split(",");
    for (String annotationFqn : annotationFqns) {
      addAnnotationFilter(scanner, annotationFqn.trim());
    }

    if (this.packages == null) {
      this.packages = "*";
    }
    if ("*".equals(this.packages)) {
      scanAndRegisterServiceClientInterfaces(scanner, "");
    } else {
      String[] basePackages = this.packages.split(",");
      for (String basePackage : basePackages) {
        scanAndRegisterServiceClientInterfaces(scanner, basePackage.trim());
      }
    }
  }

  private void scanAndRegisterServiceClientInterfaces(ClassPathScanningCandidateComponentProvider scanner,
      String basePackage) {

    for (BeanDefinition beanDefinition : scanner.findCandidateComponents(basePackage)) {
      if ((beanDefinition.getOriginatingBeanDefinition() == null) && (beanDefinition.getFactoryBeanName() == null)) {
        String beanClassName = beanDefinition.getBeanClassName();
        try {
          Class<?> beanType = Class.forName(beanClassName);
          this.beanFactory.registerSingleton(beanClassName + "-Oasp-ClientFactoryBean",
              new ClientFactoryBean<>(beanType));
          LOG.debug("Successufully registered FactoryBean for {}.", beanClassName);
        } catch (ClassNotFoundException e) {
          LOG.warn("Failed to load scanned type {}.", beanClassName, e);
        }
      }
    }
  }

  @SuppressWarnings("unchecked")
  private void addAnnotationFilter(ClassPathScanningCandidateComponentProvider scanner,
      String annotationQualifiedName) {

    try {
      Class<?> annotationType = Class.forName(annotationQualifiedName);
      if (annotationType.isAnnotation()) {
        addAnnotationFilter(scanner, (Class<? extends Annotation>) annotationType);
      } else {
        LOG.error("Expected {} to be an annotation but was not.", annotationQualifiedName);
      }
    } catch (ClassNotFoundException e) {
      LOG.warn(
          "Annotation class {} was not found. Services identified by this annotation will not be scanned and registered.",
          annotationQualifiedName, e);
    }
  }

  private void addAnnotationFilter(ClassPathScanningCandidateComponentProvider scanner,
      Class<? extends Annotation> annotationType) {

    scanner.addIncludeFilter(new AnnotationTypeFilter(annotationType));
  }

  /**
   * A {@link FactoryBean} for lazy instantiation (e.g. to prevent errors on startup for services that are never
   * injected or used).
   *
   * @param <T> type of the service API.
   */
  public class ClientFactoryBean<T> implements FactoryBean<T> {

    private final Class<T> serviceApi;

    private T client;

    /**
     * The constructor.
     *
     * @param serviceApi the service API.
     */
    public ClientFactoryBean(Class<T> serviceApi) {
      super();
      this.serviceApi = serviceApi;
    }

    @Override
    public Class<?> getObjectType() {

      return this.serviceApi;
    }

    @Override
    public T getObject() throws Exception {

      if (this.client == null) {
        this.client = ServiceClientSpringFactory.this.serviceClientFactory.create(this.serviceApi);
      }
      return this.client;
    }

    @Override
    public boolean isSingleton() {

      return true;
    }

  }

}
```

hohwille commented:
I found a new concern to consider for a general approach: CSRF
In my project we are using portal-gateway this is a reverse proxy that handles CSRF for browser clients and delegates to the actual microservices that are never accessed directly by browser (what is also technically impossible due to firewalls and network zones). Therefore these microservices intentionally do not have CSRF protection for simplicity. So we do not have problems here. However, other projects may have other setup and might require CSRF token for REST-Requests at least for non GET methods.
Hence we could have an implementation of https://github.com/hohwille/oasp4j/blob/develop/modules/service/src/main/java/io/oasp/module/service/common/api/header/ServiceHeaderCustomizer.java that would get CSRF token for according service/app from HTTP Session. If not available there, it would call a service to get the CSRF token and then store it in HTTP Session for upcoming use to prevent the cost of the additional request to get the CSRF token for multiple calls. If no HTTP Session is available (running out of servlet container) then the CSRF token is requested but not stored for subsequent calls.
However this raises a lot of questions:
* What will be the service URL to get the CSRF token? This needs to be standardized by OASP but should still be configurable.
* How do we know which services share the same CSRF token? Do we need to store CSRF token per app or even per service?

Are there projects/volunteers with demand for this? Otherwise we might just wait until this problem arises instead of creating a solution without knowing requirements and being able to do realistic tests...

amarinso commented:
Internal microservices should not concern about CSRF tokens IMHO. So the scenario you describe will be for services that are both accessed "from the outside" and "from the inside" ?

Besides that, the gateway option seems reasonable if required, so I will not consider this scenario until there is a real problem

maihacke commented:
My opinion is that hohwille oproach (validating, the token in the proxy) is best practice. It is very common, that one client access different backend services. Then it is not clear, which service to ask for the token and how to validate it in the backend. Since every microservice architecture will have something like an access proxy, api gw or what ever this is the place where csrf should be implemented.

hohwille commented:
Meanwhile, I entirely reworked the guide for this. It is available here:
https://github.com/oasp-forge/oasp4j-wiki/wiki/guide-service-client
I will also implement the tiny little improvements that came naturally whilst writing the docu.

hohwille commented:
As we are discussion the opt-in vs. opt-out strategy in OASP in general:
What should be the naming convention for spring-boot-starters from OASP4J? `oasp4j-starter-*`?
Then I should rename the modules accordingly:
* oasp4j-cxf-client to oasp4j-starter-cxf-client
* oasp4j-cxf-server to oasp4j-starter-cxf-server

anmaso commented:
-  oasp4j-starter-cxf-client
- oasp4j-starter-cxf-server


That sounds good to me 👍

hohwille commented:
I mearged with the changes that happend on develop in the forked oasp4j repo.
Hopefully I get feedback and we can get this merged before such conflicts repeat several times...

hohwille commented:
>    oasp4j-starter-cxf-client
>    oasp4j-starter-cxf-server
> That sounds good to me 👍

OK. Renamed. Done.

hohwille commented:
BTW: Integration with spring so service interface can be directly injected remains tricky. My proposal is to move this to a separate PR and let us first complete this one with the fundamentals. Any further concerns or feedback (e.g. on java packages)?

hohwille commented:
Now we also  have got SOAP support (so both JAX-WS and JAX-RS service APIs are properly supported). This makes it IMHO feature complete for dash.
Still I would like to discuss if we should split support for JAX-RS and JAX-WS (due all the dependencies) into separate starters:
* oasp4j-starter-cxf-client-rest
* oasp4j-starter-cxf-client-ws
* oasp4j-starter-cxf-server-rest
* oasp4j-starter-cxf-server-ws

WDYT?

hohwille commented:
Please vote to support spring integration: https://jira.spring.io/browse/SPR-15883

hohwille commented:
@maihacke pointed out that spring-boot-starters typically do not contain the integration/implementation code but only the dependencies and config. We should consider if we want to extract CXF implementation code from the starter modules. I think this is a good idea. If you have arguments against this (or agree), please let us know. I will implement that change if there are no reasons against it.

hohwille commented:
Thanks for your feedback. We should discuss that separate and do it in another PR.
IMHO it would even make more sense to have a separate repo "oasp4j-service" and have all the modules related to services there. Then we could also release that independently and have issues separate. But there are also various cons so we should officially discuss this in the community.
